migrations/2024-07-14-074318_enable-stopwords-in-textsearchable_index_col/up.sql

```diff
@@ -0,0 +1,45 @@
+CREATE OR REPLACE aggregate tsvector_agg (tsvector) (
+  STYPE = pg_catalog.tsvector,
+  SFUNC = pg_catalog.tsvector_concat,
+  INITCOND = ''
+);
+
+-- Add support for storing keywords considered stopwords in `crates.textsearchable_index_col` by casting
+-- to tsvector
+CREATE OR REPLACE FUNCTION trigger_crates_name_search() RETURNS trigger AS $$
+DECLARE kws tsvector;
+begin
+  SELECT
+    tsvector_agg(
+      CASE WHEN length(to_tsvector('english', keyword)) != 0 THEN to_tsvector('english', keyword)
+      ELSE (keyword || ':1')::tsvector
+      END
+      ORDER BY keyword
+  	) INTO kws
+  FROM keywords INNER JOIN crates_keywords
+    ON keywords.id = crates_keywords.keyword_id
+  WHERE crates_keywords.crate_id = new.id;
+    new.textsearchable_index_col :=
+      setweight(to_tsvector('pg_catalog.english', coalesce(new.name, '')), 'A') ||
+      setweight(kws, 'B') ||
+      setweight(to_tsvector('pg_catalog.english', coalesce(new.description, '')), 'C') ||
+      setweight(to_tsvector('pg_catalog.english', coalesce(new.readme, '')), 'D')
+  ;
+  return new;
+end
+$$ LANGUAGE plpgsql;
+
+
+-- We could update those crates with the following sql
+--
+-- WITH keywords_with_stopwords as (
+--   SELECT crate_id, keyword
+--   FROM keywords INNER JOIN crates_keywords
+--     ON id = keyword_id
+--   WHERE length(to_tsvector('english', keyword)) = 0
+-- )
+-- UPDATE crates
+-- SET updated_at = updated_at
```

Comment:
do we not have to set any other columns? 🤔

Reply:
No, because we don't want to modify the `updated_at` value. We only need to trigger `trigger_crates_tsvector_update` to update the `textsearchable_index_col` column, and I think this is sufficient.

Reply:
just to be sure, I assume you checked that this actually does not update the `updated_at` column? 😅 

(I'm a bit confused by how the automatic update is currently implemented... 🫣)

Reply:
> (I'm a bit confused by how the automatic update is currently implemented... 🫣)

Yes, this is quite opaque as it relies on a trigger behind the scenes.
We could inspect the triggers with the following sql:
```PgSql
SELECT event_object_table, action_order, trigger_name, event_manipulation
FROM information_schema.triggers
WHERE event_object_table = 'crates'
;
```
```
 event_object_table | action_order |              trigger_name              | event_manipulation 
--------------------+--------------+----------------------------------------+--------------------
 crates             |            1 | insert_crate_downloads_row             | INSERT
 crates             |            1 | trigger_crates_tsvector_update         | INSERT
 crates             |            2 | trigger_ensure_crate_name_not_reserved | INSERT
 crates             |            1 | trigger_crates_set_updated_at          | UPDATE
 crates             |            2 | trigger_crates_tsvector_update         | UPDATE
 crates             |            3 | trigger_ensure_crate_name_not_reserved | UPDATE
(6 rows)
```


> just to be sure, I assume you checked that this actually does not update the `updated_at` column? 😅

And yes, the results of the following, which updates regardless of whether the record exists or not, indicate that no `updated_at` value is currently `current_datetime`:
```console
date && psql cargo_registry <<EOF
WITH keywords_with_stopwords as (
        SELECT crate_id, keyword
        FROM keywords JOIN crates_keywords ON id = keyword_id
        WHERE length(to_tsvector('english', keyword)) = 0
), upt as (
  UPDATE crates
  SET updated_at = updated_at
  FROM keywords_with_stopwords
  WHERE id = crate_id
  -- AND NOT (keyword || ':B')::tsquery @@ textsearchable_index_col
  returning id
)
SELECT min(updated_at), max(updated_at), count(*)
FROM crates
WHERE id in (SELECT * from upt)
;

EOF
Tue Jul 23 16:07:26 CST 2024
            min             |            max             | count 
----------------------------+----------------------------+-------
 2015-12-16 00:01:49.263868 | 2024-07-13 01:40:09.733953 |   396
(1 row)
```